pkg/storegateway/bucket_index_reader.go

```diff
@@ -628,43 +631,44 @@ func (r *bucketIndexReader) preloadSeries(ctx context.Context, ids []storage.Ser
 func (r *bucketIndexReader) loadSeries(ctx context.Context, ids []storage.SeriesRef, refetch bool, start, end uint64, loaded *bucketIndexLoadedSeries, stats *safeQueryStats) error {
 	begin := time.Now()
 
-	b, err := r.block.readIndexRange(ctx, int64(start), int64(end-start))
+	reader, err := r.block.indexRangeReader(ctx, int64(start), int64(end-start))
 	if err != nil {
 		return errors.Wrap(err, "read series range")
 	}
-
-	stats.update(func(stats *queryStats) {
-		stats.seriesFetchCount++
-		stats.seriesFetched += len(ids)
-		stats.seriesFetchDurationSum += time.Since(begin)
-		stats.seriesFetchedSizeSum += int(end - start)
-	})
+	defer runutil.CloseWithLogOnErr(r.block.logger, reader, "loadSeries close range reader")
+	byteReader := &uvarintSequenceReader{r: reader, offset: start}
 
 	for i, id := range ids {
-		c := b[uint64(id)-start:]
-
-		l, n := binary.Uvarint(c)
-		if n < 1 {
-			return errors.New("reading series length failed")
+		size, err := byteReader.Uvarint(uint64(id))
+		if err != nil {
+			return err
 		}
-		if len(c) < n+int(l) {
+		seriesBytes := loaded.bytesPool.Get(int(size))
+		n, err := io.ReadFull(byteReader, seriesBytes)
+		if n != int(size) || (err != nil && errors.Is(err, io.ErrUnexpectedEOF)) {
```

Comment:
I think we can simplify this a bit. From `io.ReadFull()` doc:

> // On return, n == len(buf) if and only if err == nil.

So I think there's no need to check `n != int(size)`, but it's enough to check `err`. All in all I think it's enough to check:

```
if errors.Is(err, io.ErrUnexpectedEOF) {
```